collections/sort_by.ts

```diff
@@ -31,28 +31,35 @@ export function sortBy<T>(
     | ((el: T) => bigint)
     | ((el: T) => Date),
 ): T[] {
-  return Array.from(array).sort((a, b) => {
-    const selectedA = selector(a);
-    const selectedB = selector(b);
+  const result = array.map((value) => ({ value, selected: selector(value) }))
+    .sort((a, b) => {
+      const selectedA = a.selected;
+      const selectedB = b.selected;
 
-    if (typeof selectedA === "number") {
-      if (Number.isNaN(selectedA)) {
+      if (typeof selectedA === "number") {
+        if (Number.isNaN(selectedA)) {
+          return 1;
+        }
+
+        if (Number.isNaN(selectedB)) {
+          return -1;
+        }
+      }
+
+      if (selectedA > selectedB) {
         return 1;
       }
 
-      if (Number.isNaN(selectedB)) {
+      if (selectedA < selectedB) {
         return -1;
       }
-    }
-
-    if (selectedA > selectedB) {
-      return 1;
-    }
 
-    if (selectedA < selectedB) {
-      return -1;
-    }
+      return 0;
+    });
 
-    return 0;
-  });
+  // This does the same thing as `result.map(v=>v.value)`, but speeds it up by overwriting the original array.
+  for (let i = 0; i < result.length; i++) {
+    (result as unknown as T[])[i] = result[i].value;
+  }
+  return result as unknown as T[];
```

Comment:
Type definition is clunky, but did not come up with other methods.😓

Reply:
I think it is better if we do a `map` here. Since:
- it will gives better typing. 
- we are not returning an optimised array back to the user. I.e. If the user gives an array of integer, instead of giving back `PACKED_SMI_ELEMENTS`, we are giving back un-optimised `PACKED_ELEMENTS`. [[1]](https://v8.dev/blog/elements-kinds)
- It is cleaner, `map` is very fast, we are only sacrificing a bit of memory here.

Reply:
Certainly, it seems that there is not much difference in execution time. I will change this to use `map`.